components/base/sensorcontrolled/movestraight.go

```diff
@@ -0,0 +1,204 @@
+)
+
+const (
+	slowDownDistGain      = .1
+	maxSlowDownDist       = 100 // mm
+	moveStraightErrTarget = 20  // mm
+	headingGain           = 1.
+)
+
+func (sb *sensorBase) MoveStraight(
+	ctx context.Context, distanceMm int, mmPerSec float64, extra map[string]interface{},
+) error {
+	ctx, done := sb.opMgr.New(ctx)
+	defer done()
+	sb.setPolling(false)
+	straightTimeEst := time.Duration(int(time.Second) * int(math.Abs(float64(distanceMm)/mmPerSec)))
+	startTime := time.Now()
+	timeOut := 5 * straightTimeEst
+	if timeOut < 10*time.Second {
+		timeOut = 10 * time.Second
+	}
+
+	if sb.position == nil || len(sb.controlLoopConfig.Blocks) == 0 {
+		sb.logger.CWarnf(ctx,
+			"Position reporting sensor not available, and no control loop is configured, using base %s MoveStraight",
+			sb.controlledBase.Name().ShortName())
+		sb.stopLoop()
+		return sb.controlledBase.MoveStraight(ctx, distanceMm, mmPerSec, extra)
+	}
+
+	initialHeading, err := sb.headingFunc(ctx)
+	if err != nil {
+		return err
+	}
+	// make sure the control loop is enabled
+	if sb.loop == nil {
+		if err := sb.startControlLoop(); err != nil {
+			return err
+		}
+	}
+
+	// initialize relevant parameters for moving straight
+	slowDownDist := calcSlowDownDist(distanceMm)
+
+	var initPos *geo.Point
+
+	if sb.position != nil {
+		initPos, _, err = sb.position.Position(ctx, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	ticker := time.NewTicker(time.Duration(1000./sb.controlLoopConfig.Frequency) * time.Millisecond)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			// do not return context canceled errors, just log them
+			if errors.Is(ctx.Err(), context.Canceled) {
+				sb.logger.Error(ctx.Err())
+				return nil
+			}
+			return ctx.Err()
+		case <-ticker.C:
+			var errDist float64
+
+			angVelDes, err := sb.calcHeadingControl(ctx, initialHeading)
+			if err != nil {
+				return err
+			}
+
+			if sb.position != nil {
+				errDist, err = sb.calcPositionError(ctx, distanceMm, initPos)
+				if err != nil {
+					return err
+				}
+			}
+
+			if errDist < moveStraightErrTarget {
+				return sb.Stop(ctx, nil)
+			}
+
+			linVelDes := calcLinVel(errDist, mmPerSec, slowDownDist)
+			if err != nil {
+				return err
+			}
+
+			// update velocity controller
+			if err := sb.updateControlConfig(ctx, linVelDes/1000.0, angVelDes); err != nil {
+				return err
+			}
+
+			// exit if the straight takes too long
+			if time.Since(startTime) > timeOut {
+				sb.logger.CWarn(ctx, "exceeded time for MoveStraightCall, stopping base")
+
+				return sb.Stop(ctx, nil)
+			}
+		}
+	}
+}
+
+// calculate the desired angular velocity to correct the heading of the base.
+func (sb *sensorBase) calcHeadingControl(ctx context.Context, initHeading float64) (float64, error) {
+	currHeading, err := sb.headingFunc(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	headingErr := initHeading - currHeading
+	headingErrWrapped := headingErr - (math.Floor((headingErr+180.)/(2*180.)))*2*180. // [-180;180)
```

Comment:
wrap around math always confuses me... do you have a unit test for this? could be helpful if you haven't seen this scenario with hardware testing

Reply:
I can add a test for this yeah. I'll validate that the wrapped angle error falls within +-180 degrees for a range of setpoints(initHeadings in this case)